### PR TITLE
OPENSRC-105 Make sure all the promises will reject in case of internal throw

### PR DESCRIPTION
**Pull request by lastperson**

### CODE REVIEW

**Comment by orlandoc01** [comment]
Just wanted to offer an alternative strategy for this PR that may be worth considering. @imolfar and I had discussed how using `Promise.promisify` may be used as a means of promisifying some of the Registry callback functions but opted against it because it looked like we would have to change a lot of the methods that were constructed and we wanted to keep the code styling coherent. However, this PR looks like you're doing that already so it may be worthwhile to reconsider this tactic. I'll briefly describe an example implementation and the reasoning behind it and leave it up you guys whether or not to adopt this strategy.

The issue with the way Promises are currently constructed in this library is that it is an anti-pattern because the registry method callback API automatically support Promisification out-of-the-box, so doing it manually using a Promise constructor with `resolve` and reject` leads to more code and is more likely to lead to uncaught errors. (See http://bluebirdjs.com/docs/anti-patterns.html example 2 for more info). So, as an alternative, you could create a function which wraps the function in a `Promise.promisify` call and then immediately invokes it and use this function instead of the `catcher` function you implemented above.

For example, say you have a helper function like:
```const P = (context, func, ...args) => Promise.promisify(func.bind(context, ...args))();```

You could then take code like this: 
```Consumer.prototype.getRegistrant = function(address) {
    var self = this;
 +  return catcher(function(fulfill, reject) {
      self.registrar.registrantIndex.call(address, {from: self.address}, function(err, index) {
        if (err) {
          console.error(err);
         reject(err);
         return;
       }
       fulfill(self.getRegistrantbyIndex(index));
     });
   });
 };
 ```
and turn it into something like this:
```
Consumer.prototype.getRegistrant = function(address) {
  return P(this, this.registrar.registrantIndex.call, address, {from: this.address})
  .then(index => this.getRegistrantbyIndex(index));
};
```
Like I said though, just a suggestion for an alternative way of handling this issue. I wouldn't suggest it if it was only one function that needed this wrapping but since it looks like you're changing most of them, it may be worthwhile to then just invest the time and promisify them now instead. I've tested this code locally on some of the methods just out of curiosity and it works fine. If you opted for this, you would then just add that helper function or something like that and replace each `catcher` call with that helper function instead and get rid of the callbacks.

**Comment by lastperson** [comment]
@orlandoc01 @imolfar please review new diff.

File: lib/consumer.js [hunk 1]
```diff
@@ -64,15 +54,8 @@ Consumer.prototype.getRegistrar = function() {
...
-  return new Promise(function(fulfill, reject) {
-    self.registrar.registrantIndex.call(address, {from: self.address}, function(err, index) {
-      if (err) {
-        console.error(err);
-        reject(err);
-        return;
-      }
-      fulfill(self.getRegistrantbyIndex(index));
-    });
+  return self.registrar.registrantIndexAsync(address, {from: self.address}).then(function(index) {
```
Comment (orlandoc01) [inline lib/consumer.js]: You can make this even shorter by getting rid of the `self` variable. I also like to have my `then blocks start on a new line and you can use the arrow functions to remove boilerplate. So something like
```
return this.registrar.registrantIndexAsync(address, {from: this.addresss})
.then(index => this.getRegistrantbyIndex(index));
```

**Comment by lastperson** [comment]
I agree, though refactoring can be done over and over again and it is out of scope of this task.

File: lib/consumer.js [hunk 2]
```diff
@@ -116,56 +89,48 @@ Consumer.prototype.getRegistrarAddressOnRegistry = function() {
...
-      var encodedIdentities = data[0];
-      if (encodedIdentities.length == 0) {
-        reject('Thing is not existing');
-        return;
-      }
+  return Promise.try(function() {
+    return OrUtils.urn.packer.encodeAndChunk(OrUtils.common.compressAll(identity));
+  }).then(function(encodedIdentity) {
+    return self.registry.getThingAsync(encodedIdentity, {from: self.address});
+  }).then(function(data) {
```
Comment (orlandoc01) [inline lib/consumer.js]: Similiar style comment here. If you use arrow functions you can get rid of the `self` declaration at the top. Whether or not these are one line functions is entirely up to you. Something like
```
return Promise.try(() => OrUtils.urnpacker.encodeAndCHunc(OrbUtils.common.compressAll(identity)))
.then(id => this.registry.getThingsAsync(id, {from: this.address})
.then(data =>
  ...
```

File: lib/consumer.js [hunk 3]
```diff
@@ -176,19 +141,11 @@ Consumer.prototype.getThing = function(identity) {
...
-        if (proto.slice(0, 2) == '0x') {
-          proto = proto.slice(2);
-        }
-        fulfill(Schema.decodeHex(proto));
-    });
+  return this.registry.schemasAsync(index, {from: this.address}).then(function(proto) {
+    if (proto.slice(0, 2) == '0x') {
+      proto = proto.slice(2);
+    }
+    return Schema.decodeHex(proto);
```
Comment (orlandoc01) [inline lib/consumer.js]: Same style comments about `then` blocks and arrow functions. You could also use a ternary operator here to condense it a bit more
```
return this.registry.schemasAsync(index, {from: this.address})
.then(proto => Schema.decodeHex(proto.includes('0x') ? proto.slice(2) : proto))
```

Comment (lastperson) [inline lib/consumer.js]: I know the es6 goodies, but the whole sdk is not following es6, so again it is out of scope of this task.

File: lib/consumer.js [hunk 4]
```diff
@@ -199,16 +156,7 @@ Consumer.prototype.getSchema = function(index) {
...
-    self.registry.getSchemasLenght.call({from: self.address}, function(error, length) {
-      if (error) {
-        console.error(error);
-        reject(error);
-        return;
-      }
-      fulfill(length);
-    });
-  }).then(function(length) {
+  return self.registry.getSchemasLenghtAsync({from: self.address}).then(function(length) {
```
Comment (orlandoc01) [inline lib/consumer.js]: Same style comments with using `self` and arrow functions and `then` block as before. Also, the `getSchemasLenghtAsync` method has length misspelled. Is this something we control or is this in the library?

File: lib/consumer.js [hunk 5]
```diff
@@ -225,16 +173,9 @@ Consumer.prototype.getSchemas = function() {
...
-    self.registrar.registrants.call(index, function(error, data) {
-      if (error) {
-        console.error(error);
-        reject(error);
-        return;
-      }
-      data[1] = self.decodeRegistrant(data[1]);
-      fulfill(data);
-    });
+  return self.registrar.registrantsAsync(index).then(function(data) {
```
Comment (orlandoc01) [inline lib/consumer.js]: Styling for `this` and arrow functions as explained in above comments

File: lib/consumer.js [hunk 6]
```diff
@@ -269,48 +210,18 @@ Consumer.prototype.verifyIdentity = function(identityURN, message, signature) {
...
-        }
-        resolve(deletions);
-      });
-    })
-  ]).then(function(results) {
-    var events = results[0].concat(results[1]).concat(results[2]);
-    var sortedEvents = sortEvents(events);
+  var filter = Promise.promisifyAll(this.registry.allEvents({fromBlock: fromBlock}));
+  return filter.getAsync().then(function(events) {
+    var filteredEvents = events.filter(function(event) {
```
Comment (orlandoc01) [inline lib/consumer.js]: Styling with arrow functions and `then block`. Could also rewrite filter like
```
let filteredEvents = events.filter(event => event.event !== 'Error');
```

File: lib/consumer.js [hunk 7]
```diff
@@ -371,36 +282,18 @@ Consumer.prototype.getThings = function(onlyValid, fromBlock, finalCallback) {
...
-        }
-        resolve(alternations);
-      });
-    })
-  ]).then(function(results) {
-    var events = results[0].concat(results[1]);
-    var sortedEvents = sortEvents(events);
+  var filter = Promise.promisifyAll(self.registrar.allEvents({fromBlock: fromBlock}));
+  return filter.getAsync().then(function(events) {
+    var filteredEvents = events.filter(function(event) {
```
Comment (orlandoc01) [inline lib/consumer.js]: Same comments as above

File: lib/provider.js [hunk 8]
```diff
@@ -157,7 +158,7 @@ function Provider(rpcUrl, userType, secret, contracts, finalCallback) {
 
   engine.start();
 
-  this.web3 = web3;
+  this.web3 = Promise.promisifyAll(web3);
```
Comment (orlandoc01) [inline lib/provider.js]: 👍

File: lib/registrant.js [hunk 9]
```diff
@@ -50,43 +50,24 @@ Registrant.prototype.createThing = function(thing, schemaIndex) {
...
-      }
-
-      if (protoBufferSchema.slice(0, 2) == '0x') {
-        protoBufferSchema = protoBufferSchema.slice(2);
-      }
-      var builder = ProtoBuf.loadProto(Schema.decodeHex(protoBufferSchema).get('definition'));
-      var Thing = builder.build("Thing");
-      var dataProto = new Thing(thing.data);
-      var dataChunks = OrUtils.bytes32Array.slice(dataProto.encodeHex());
+  return self.registry.schemasAsync(schemaIndex, {from: self.address}).then(function(protoBufferSchema) {
```
Comment (orlandoc01) [inline lib/registrant.js]: same comments with `self` and arrow functions as above

File: lib/registrant.js [hunk 10]
```diff
@@ -105,51 +86,34 @@ Registrant.prototype.createThings = function(things, schemaIndex) {
...
-      }
-      // Pack data into chunked hex bytes32 packets using ProtoBuffer
-      var builder = ProtoBuf.loadProto(Schema.decodeHex(protoBufferSchema).get('definition'));
-      var Thing = builder.build("Thing");
-
-      var idsChunks = [];
-      var idsPerThing = [];
-      var dataLength = [];
-      var dataChunks = [];
+  return self.registry.schemasAsync(schemaIndex, {from: self.address}).then(function(protoBufferSchema) {
```
Comment (orlandoc01) [inline lib/registrant.js]: same comments with `self`, then blocks, and arrow functions as above

File: lib/registrant.js [hunk 11]
```diff
@@ -238,16 +178,9 @@ Registrant.prototype.addIdentities = function(identity, newIdentities) {
...
     var encodedIdentity = OrUtils.urn.packer.encodeAndChunk(OrUtils.common.compressAll(identity));
-    self.registry.setThingValid(encodedIdentity, isValid, {from: self.address, gas: 2000000}, function(err, data) {
-      if (err) {
-        console.error(err);
-        reject(err);
-        return;
-      }
-      fulfill(data);
-    });
+    return self.registry.setThingValidAsync(encodedIdentity, isValid, {from: self.address, gas: 2000000});
```
Comment (orlandoc01) [inline lib/registrant.js]: Same comments with `self` and arrow functions as above

**Comment by orlandoc01** [comment]
Great job addressing this anti-pattern using `Promisify.all`. It's a pretty clean solution and even better then the one I originally suggested in my post. Nice work! :+1:. Just made a few minor styling suggestions that deal with new features in ES6 that could save us even a few more lines of code and make it more consise and understandable. This mainly deals with taking advantage of the new lexical `this` binding in arrow functions. Feel free to ignore these style suggestions if you prefer, but there is one issue that should be addressed with the misspelling of lenght in 'lib/consumer.js` on line. Otherwise, you're good to go!

**Comment by lastperson** [comment]
@orlandoc01 
> one issue that should be addressed with the misspelling of lenght in lib/consumer.js on line 159
It should be fixed inside the Registry contract first, please create separate task.

**Pull request merged**